File: scala/src/main/scala/domain/Bank.scala
package domain

case class Bank(exchangeRates: Map[String, Double] = Map.empty) {
  private def keyFor(from: String, to: String): String = s"$from->$to"

  def addExchangeRate(from: String, to: String, rate: Double): Bank =
    Bank(exchangeRates + (keyFor(from, to) -> rate))

  def convert(money: Money, currency: String): Either[String, Money] =
    if (canConvert(money.currency, currency)) convertSafely(money, currency)
    else Left(keyFor(money.currency, currency))

  private def canConvert(from: String, to: String): Boolean =
    from == to || exchangeRates.contains(keyFor(from, to))

  private def convertSafely(
      money: Money,
      currency: String
  ): Either[String, Money] =
    Right(
      if (currency == money.currency) money
      else
        Money(
          money.amount * exchangeRates(keyFor(money.currency, currency)),
          currency
        )
    )
}

File: c#/money-problem.Domain/Bank.cs
using System.Collections.Immutable;

namespace money_problem.Domain
{
    public record Bank
    {
        private readonly ImmutableDictionary<string, double> _exchangeRates;

        private Bank(ImmutableDictionary<string, double> exchangeRates) => _exchangeRates = exchangeRates;

        public static Bank WithExchangeRate(Currency from, Currency to, double rate) =>
            new Bank(new Dictionary<string, double>().ToImmutableDictionary())
                .AddExchangeRate(from, to, rate);

        public Bank AddExchangeRate(Currency from, Currency to, double rate) =>
            new(_exchangeRates.Add(KeyFor(from, to), rate));

        private static string KeyFor(Currency from, Currency to) => $"{from}->{to}";

        public Money Convert(Money money, Currency currency)
        {
            CheckExchangeRates(money.Currency, currency);
            return currency == money.Currency
                ? money
                : new Money(money.Amount * _exchangeRates[KeyFor(money.Currency, currency)], currency);
        }

        private void CheckExchangeRates(Currency from, Currency to)
        {
            if (from == to) return;
            var key = KeyFor(from, to);
            if (!_exchangeRates.ContainsKey(key))
                throw new MissingExchangeRateException(key);
        }
    }
}